src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs

```diff
@@ -737,23 +739,70 @@ public static unsafe void CheckArrayStore(object array, object obj)
             throw array.GetMethodTable()->GetClasslibException(ExceptionIDs.ArrayTypeMismatch);
         }
 
-        internal struct ArrayElement
+        private static unsafe void ThrowIndexOutOfRangeException(object?[] array)
         {
-            public object Value;
+            // Throw the index out of range exception defined by the classlib, using the input array's MethodTable*
+            // to find the correct classlib.
+            throw array.GetMethodTable()->GetClasslibException(ExceptionIDs.IndexOutOfRange);
+        }
+
+        private static unsafe void ThrowArrayMismatchException(object?[] array)
+        {
+            // Throw the array type mismatch exception defined by the classlib, using the input array's MethodTable*
+            // to find the correct classlib.
+            throw array.GetMethodTable()->GetClasslibException(ExceptionIDs.ArrayTypeMismatch);
         }
 
         //
         // Array stelem/ldelema helpers with RyuJIT conventions
         //
+
+        [RuntimeExport("RhpLdelemaRef")]
+        public static unsafe ref object? LdelemaRef(object?[]? array, nint index, IntPtr elementType)
+        {
+            Debug.Assert(array is null || array.GetMethodTable()->IsArray, "first argument must be an array");
+
+#if INPLACE_RUNTIME
+            if ((nuint)index >= (uint)array.Length)
+                ThrowIndexOutOfRangeException(array);
```

Comment:
This can just use `ThrowHelper.ThrowIndexOutOfRangeException` given that is for `INPLACE_RUNTIME`. It is unnecessary to go through `GetClasslibException` indirection for `INPLACE_RUNTIME`. The main point of `INPLACE_RUNTIME` is to avoid these indirections.

Reply:
I wanted to remove the `!INPLACE_RUNTIME` code in the next PR and to keep this like earlier for now.
Also I think that `ThrowHelper` is not in the `Test.CoreLib` (not sure what that is for tbh but right now there's a build failure here due to `MemoryMarshal` not being there too).

Reply:
Test.CoreLib is very minimal implementation of CoreLib for testing, debugging and various experiments. It is fine to add minimal implementations of MemoryMarshal or ThrowHelper to it to keep it working.